### ISSUE TITLE
Enabling In Game Vsync on Unity-based games significantly decreases performance

### ISSUE THREAD

**Issue by Thread6376** [doitsujin/dxvk#3654]
Enabling vsync on Unity-based games significantly decreases performance and results in low GPU utilization. Disabling vsync improves game speed and increases GPU utilization. To address this issue, you can disable in-game vsync and add `dxgi.syncInterval = 1` to dxvk.conf to enable vsync without sacrificing performance. However, not all Unity-based games offer the option to enable or disable vsync, and some games still limit the FPS even when vsync is disabled.

When using gamescope and enabling `ENABLE_GAMESCOPE_WSI=1` with vsync on (both in-game and in dxvk.conf), the game runs at 72 FPS, which is half of my monitor's refresh rate.

This problem occurs both on Nvidia and AMD GPU.

### Software information
Tested on some Unity-based games with vsync on and off, like:
- TUNIC
- Hollow Knight
- Death's Door
- GRIME
- A Short Hike
- And many more

### System information
- GPU: Ryzen 5 4600H + GTX 1650 Ti
- Driver: 535.104.05
- Wine version: soda-7.0-9
- DXVK version: 2.3

### Apitrace file(s)
https://anonfiles.me/yJr/seaofstarstrace.7z

**Comment by doitsujin** [comment]
I can't confirm this at all, the Sea of Stars demo runs at a locked 144 FPS on my AMD desktop with Vsync on, and with a locked 60 FPS on my Nvidia laptop, which is expected. This is using Proton 8.0-3.

No idea what soda-7.0 is but please try this on a proper up-to-date Proton setup.

**Comment by doitsujin** [comment]
Actually this game uses `GetFrameStatistics` when vsync is enabled, I missed that earlier.

For this to work properly you will need a setup that implements present_wait, so yeah, your wine build is most likely the issue here.

**Comment by Thread6376** [comment]
Soda is bottles wine. I had this problem for years. I use laptop with amd + nvidia. Tested on X11 and wayland with wine-ge amd soda with clean bottles (havent tried steam). All unity based game have bad performance when enabling vsync, but after disabling it , it ran flawlessly.

Maybe trying with sea of stars is not the best to reproduce this problem. Game like tunic or grime, which is a bit on the heavy side, is better to test this problem.

Btw, Sea of stars capped the fps to 60 when disabling vsync.

**Comment by doitsujin** [comment]
The whole point I'm  making is is that weird old wine builds don't necessarily support the Vulkan extensions needed for certain functionality to work, so **please** try this in a known-good environment.

Anyway, still can't reproduce here.

**Comment by Thread6376** [comment]
> The whole point I'm making is is that weird old wine builds don't necessarily support the Vulkan extensions needed for certain functionality to work, so **please** try this in a known-good environment.
> 
> Anyway, still can't reproduce here.

WDYM by a known-good environment? I've tested also Wine-GE-Proton8-14 and latest nvidia using bottles. Isn't it good?

I'm currently downloading TUNIC to test and will post some screenshot and video.

**Comment by doitsujin** [comment]
Yeah, don't know what to tell you, it works fine on my Nvidia laptop at least with X11 (don't have a working Wayland setup on this thing). With an external display as well, although annoyingly the APU can only drive it at 100Hz over HDMI for some stupid reason so I can't test higher refresh rates.

**Comment by Thread6376** [comment]
Well it's too bad, now I'm think I'm the only person with this problem in the world. At least for me, this problem exists with or without external monitor. Maybe my gpu have bad performance so i could reproduce it. Maybe your gpu is too good even the vsync problem couldnt make it worst.

Anyways there's always the workaround. And Sea of Stars shouldn't have capped the game at 60fps with vsync off and the workaround render useless.

**Comment by Iglu47** [comment]
@Thread6376 what kernel version are you using? Do you have SMT enabled? If so, could you tell us how the environment variable `WINE_CPU_TOPOLOGY=6:0,2,4,6,8,10` (affects Wine builds based on Proton8) affects Unity games at your setup?

**Comment by Thread6376** [comment]
> @Thread6376 what kernel version are you using? 
6.4.13. This problem is not cause by kernel.

> Do you have SMT enabled? If so, could you tell us how the environment variable `WINE_CPU_TOPOLOGY=6:0,2,4,6,8,10` (affects Wine builds based on Proton8) affects Unity games at your setup?
> 
My asus laptop doesn't allow to disable SMT in the bios, heck there aren't any useful option. The env made my game ran worst.

**Comment by Sid127** [comment]
for turning off SMT, could try it kernel side with `echo off | sudo tee /sys/devices/system/cpu/smt/control`

**Comment by netborg-afps** [comment]
I just have noticed you are using dxvk 2.2, whereas in the description you told us you would use 2.3.

**Comment by Thread6376** [comment]
With smt off and the env, the game still ran terribly compared to smt on and without the env. The vsync bug is still there.

**Comment by Thread6376** [comment]
> I just have noticed you are using dxvk 2.2, whereas in the description you told us you would use 2.3.

Tried it with both, still have the same problem. I have this bug since fedora 34 and didn't report it since there's the workaround I mentioned before.

**Comment by netborg-afps** [comment]
It's not clear whether this is a bug?! 
It could very well be this to be intentional by Unity?
To me this sounds like your system isn't fast enough to complete every frame in under 7 ms.
It might just be that using the game option v-sync it waits for v-blank while it doesn't using `dxgi.syncInterval = 1`.

**Comment by Thread6376** [comment]
> It's not clear whether this is a bug?! 
>
If you see the screenshot and video that I gave there's a difference. I've tried multiple unity games and encounters the same problem with vsync.

> It could very well be this to be intentional by Unity?
>
This problem doesn't exists on windows or other engine, just unity. I think it's a bug on linux, I just don't know whose fault it is.

Then again some games have bad vsync implementation and on windows I always disable in game vsync and enable Nvidia Control Panel Vsync.

I don't think it's intentional because I haven't seen vsync impacts performance this severe.

> To me this sounds like your system isn't fast enough to complete every frame in under 7 ms.
>
This only happens on Unity based games  on linux and I don't think it's because of my low end system.

**Comment by netborg-afps** [comment]
Do you have to possibility to turn on adaptive sync? Maybe the "wait for v-blank" (which is my current hypothesis, not a fact...) would be handled differently then?

**Comment by Thread6376** [comment]
> Do you have to possibility to turn on adaptive sync? Maybe the "wait for v-blank" (which is my current hypothesis, not a fact...) would be handled differently then?

Vrr is turned on

**Comment by netborg-afps** [comment]
I checked out Cities Skylines and have same fps with v-sync and without, cannot reproduce.
